File: implementations/typescript/src/conformance.test.ts
import { describe, it, expect } from "vitest";
import { parse, stringify } from "./index.js";
import type { RDNValue, RDNTimeOnly, RDNDuration } from "./types.js";
import { readFileSync, readdirSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const SUITE_DIR = join(__dirname, "../../../test-suite");

// ── Normalize parsed RDN values to $type-tagged JSON for comparison ─────

function normalizeForComparison(value: RDNValue): unknown {
  if (value === null) return null;
  if (typeof value === "string" || typeof value === "boolean") return value;

  if (typeof value === "number") {
    if (Number.isNaN(value)) return { $type: "Number", value: "NaN" };
    if (value === Infinity) return { $type: "Number", value: "Infinity" };
    if (value === -Infinity) return { $type: "Number", value: "-Infinity" };
    return value;
  }

  if (typeof value === "bigint") {
    return { $type: "BigInt", value: String(value) };
  }

  if (value instanceof Date) {
    return { $type: "Date", value: value.toISOString() };
  }

  if (value instanceof RegExp) {
    return { $type: "RegExp", value: { source: value.source, flags: value.flags } };
  }

  if (value instanceof Uint8Array) {
    // Encode to base64 for comparison
    return { $type: "Binary", value: encodeBase64(value) };
  }

  if (value instanceof Map) {
    const entries: unknown[] = [];
    for (const [k, v] of value) {
      entries.push([normalizeForComparison(k as RDNValue), normalizeForComparison(v as RDNValue)]);
    }
    return { $type: "Map", value: entries };
  }

  if (value instanceof Set) {
    const elements: unknown[] = [];
    for (const v of value) {
      elements.push(normalizeForComparison(v as RDNValue));
    }
    return { $type: "Set", value: elements };
  }

  if (typeof value === "object" && "__type__" in value) {
    if (value.__type__ === "TimeOnly") {
      const t = value as RDNTimeOnly;
      return { $type: "TimeOnly", value: { hours: t.hours, minutes: t.minutes, seconds: t.seconds, milliseconds: t.milliseconds } };
    }
    if (value.__type__ === "Duration") {
      return { $type: "Duration", value: (value as RDNDuration).iso };
    }
  }

  if (Array.isArray(value)) {
    return value.map(v => normalizeForComparison(v as RDNValue));
  }

  // Plain object
  const obj = value as { [key: string]: RDNValue };
  const result: Record<string, unknown> = {};
  for (const key of Object.keys(obj)) {
    result[key] = normalizeForComparison(obj[key]!);
  }
  return result;
}

function encodeBase64(bytes: Uint8Array): string {
  const charset = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  let result = "";
  const len = bytes.length;
  let i = 0;
  for (; i + 2 < len; i += 3) {
    const a = bytes[i]!, b = bytes[i + 1]!, c = bytes[i + 2]!;
    result += charset[a >> 2]! + charset[((a & 0x03) << 4) | (b >> 4)]! + charset[((b & 0x0F) << 2) | (c >> 6)]! + charset[c & 0x3F]!;
  }
  if (i < len) {
    const a = bytes[i]!;
    if (i + 1 < len) {
      const b = bytes[i + 1]!;
      result += charset[a >> 2]! + charset[((a & 0x03) << 4) | (b >> 4)]! + charset[(b & 0x0F) << 2]! + "=";
    } else {
      result += charset[a >> 2]! + charset[(a & 0x03) << 4]! + "==";
    }
  }
  return result;
}

function deepEqual(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (a === null || b === null) return false;
  if (typeof a !== typeof b) return false;
  if (typeof a !== "object") return false;

  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    return a.every((v, i) => deepEqual(v, b[i]));
  }

  const aKeys = Object.keys(a as object);
  const bKeys = Object.keys(b as object);
  if (aKeys.length !== bKeys.length) return false;
  return aKeys.every(k => deepEqual((a as Record<string, unknown>)[k], (b as Record<string, unknown>)[k]));
}

// ── Valid tests ─────────────────────────────────────────────────────────

describe("Conformance: valid", () => {
  const validDir = join(SUITE_DIR, "valid");
  let files: string[];
  try {
    files = readdirSync(validDir).filter((f: string) => f.endsWith(".rdn"));
  } catch {
    files = [];
  }

  for (const file of files) {
    const name = file.replace(".rdn", "");
    it(`valid/${name}`, () => {
      const input = readFileSync(join(validDir, file), "utf-8");
      const expectedJson = readFileSync(join(validDir, `${name}.expected.json`), "utf-8");
      const expected = JSON.parse(expectedJson);
      const parsed = parse(input);
      const normalized = normalizeForComparison(parsed);
      expect(deepEqual(normalized, expected)).toBe(true);
    });
  }
});

// ── Invalid tests ───────────────────────────────────────────────────────

describe("Conformance: invalid", () => {
  const invalidDir = join(SUITE_DIR, "invalid");
  let files: string[];
  try {
    files = readdirSync(invalidDir).filter((f: string) => f.endsWith(".rdn"));
  } catch {
    files = [];
  }

  for (const file of files) {
    const name = file.replace(".rdn", "");
    it(`invalid/${name}`, () => {
      const input = readFileSync(join(invalidDir, file), "utf-8");
      expect(() => parse(input)).toThrow();
    });
  }
});

// ── Roundtrip tests ─────────────────────────────────────────────────────

describe("Conformance: roundtrip", () => {
  const roundtripDir = join(SUITE_DIR, "roundtrip");
  let files: string[];
  try {
    files = readdirSync(roundtripDir).filter((f: string) => f.endsWith(".rdn"));
  } catch {
    files = [];
  }

  for (const file of files) {
    const name = file.replace(".rdn", "");
    it(`roundtrip/${name}`, () => {
      const input = readFileSync(join(roundtripDir, file), "utf-8");
      const parsed1 = parse(input);
      const serialized = stringify(parsed1);
      expect(serialized).toBeDefined();
      const parsed2 = parse(serialized!);
      const norm1 = normalizeForComparison(parsed1);
      const norm2 = normalizeForComparison(parsed2);
      expect(deepEqual(norm1, norm2)).toBe(true);
    });
  }
});

File: packages/rdn-go/conformance_test.go
package rdn

import (
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

const testSuiteDir = "../../test-suite"

// valueToJSON converts an rdn.Value to a Go any suitable for JSON comparison.
// Extended types use the $type tagged convention from the conformance suite.
func valueToJSON(v Value) any {
	switch v.kind {
	case KindNull:
		return nil
	case KindBool:
		return v.boolean
	case KindNumber:
		if math.IsNaN(v.num) {
			return map[string]any{"$type": "Number", "value": "NaN"}
		}
		if math.IsInf(v.num, 1) {
			return map[string]any{"$type": "Number", "value": "Infinity"}
		}
		if math.IsInf(v.num, -1) {
			return map[string]any{"$type": "Number", "value": "-Infinity"}
		}
		// Return as float64 (JSON number)
		return v.num
	case KindBigInt:
		return map[string]any{"$type": "BigInt", "value": v.str}
	case KindString:
		return v.str
	case KindArray:
		elems := v.Array()
		arr := make([]any, len(elems))
		for i, elem := range elems {
			arr[i] = valueToJSON(elem)
		}
		return arr
	case KindObject:
		pairs := v.Object()
		m := make(map[string]any, len(pairs))
		for _, kv := range pairs {
			m[kv.Key] = valueToJSON(kv.Value)
		}
		return m
	case KindDateTime:
		t := v.Time().UTC()
		return map[string]any{"$type": "Date", "value": t.Format("2006-01-02T15:04:05.000Z")}
	case KindTimeOnly:
		to := v.TimeOnlyValue()
		return map[string]any{
			"$type": "TimeOnly",
			"value": map[string]any{
				"hours": float64(to.Hours), "minutes": float64(to.Minutes),
				"seconds": float64(to.Seconds), "milliseconds": float64(to.Milliseconds),
			},
		}
	case KindDuration:
		return map[string]any{"$type": "Duration", "value": v.str}
	case KindRegExp:
		re := v.RegExpValue()
		return map[string]any{
			"$type": "RegExp",
			"value": map[string]any{"source": re.Source, "flags": re.Flags},
		}
	case KindBinary:
		// Encode as base64 for comparison
		enc := newEncoder("", "")
		defer putEncodeState(enc.buf)
		enc.encodeBase64(v.Bytes())
		raw := enc.buf.String()
		// Strip the b"..." wrapper
		b64str := raw[2 : len(raw)-1]
		return map[string]any{"$type": "Binary", "value": b64str}
	case KindMap:
		mapEntries := v.Map()
		entries := make([]any, len(mapEntries))
		for i, entry := range mapEntries {
			entries[i] = []any{valueToJSON(entry.Key), valueToJSON(entry.Value)}
		}
		return map[string]any{"$type": "Map", "value": entries}
	case KindSet:
		elems := v.Array()
		items := make([]any, len(elems))
		for i, elem := range elems {
			items[i] = valueToJSON(elem)
		}
		return map[string]any{"$type": "Set", "value": items}
	case KindTuple:
		elems := v.Array()
		arr := make([]any, len(elems))
		for i, elem := range elems {
			arr[i] = valueToJSON(elem)
		}
		return arr
	default:
		return nil
	}
}

func TestConformanceValid(t *testing.T) {
	validDir := filepath.Join(testSuiteDir, "valid")
	entries, err := os.ReadDir(validDir)
	if err != nil {
		t.Fatalf("Cannot read valid test directory: %v", err)
	}

	for _, entry := range entries {
		name := entry.Name()
		if !strings.HasSuffix(name, ".rdn") {
			continue
		}
		baseName := strings.TrimSuffix(name, ".rdn")
		t.Run(baseName, func(t *testing.T) {
			rdnPath := filepath.Join(validDir, name)
			expectedPath := filepath.Join(validDir, baseName+".expected.json")

			rdnData, err := os.ReadFile(rdnPath)
			if err != nil {
				t.Fatalf("Cannot read RDN file: %v", err)
			}
			expectedData, err := os.ReadFile(expectedPath)
			if err != nil {
				t.Fatalf("Cannot read expected JSON file: %v", err)
			}

			// Parse RDN
			val, err := Parse(rdnData)
			if err != nil {
				t.Fatalf("Parse failed: %v", err)
			}

			// Convert to JSON-compatible structure
			got := valueToJSON(val)

			// Parse expected JSON
			var expected any
			if err := json.Unmarshal(expectedData, &expected); err != nil {
				t.Fatalf("Cannot parse expected JSON: %v", err)
			}

			// Compare by re-serializing both to JSON
			gotJSON, err := json.Marshal(got)
			if err != nil {
				t.Fatalf("Cannot marshal got: %v", err)
			}
			expectedJSON, err := json.Marshal(expected)
			if err != nil {
				t.Fatalf("Cannot marshal expected: %v", err)
			}

			if string(gotJSON) != string(expectedJSON) {
				t.Errorf("Mismatch:\n  got:      %s\n  expected: %s", gotJSON, expectedJSON)
			}
		})
	}
}

func TestConformanceInvalid(t *testing.T) {
	invalidDir := filepath.Join(testSuiteDir, "invalid")
	entries, err := os.ReadDir(invalidDir)
	if err != nil {
		t.Fatalf("Cannot read invalid test directory: %v", err)
	}

	for _, entry := range entries {
		name := entry.Name()
		if !strings.HasSuffix(name, ".rdn") {
			continue
		}
		baseName := strings.TrimSuffix(name, ".rdn")
		t.Run(baseName, func(t *testing.T) {
			rdnPath := filepath.Join(invalidDir, name)
			data, err := os.ReadFile(rdnPath)
			if err != nil {
				t.Fatalf("Cannot read file: %v", err)
			}

			_, err = Parse(data)
			if err == nil {
				t.Errorf("Expected parse error for %s but got success", name)
			}
		})
	}
}

func TestConformanceRoundtrip(t *testing.T) {
	roundtripDir := filepath.Join(testSuiteDir, "roundtrip")
	entries, err := os.ReadDir(roundtripDir)
	if err != nil {
		t.Fatalf("Cannot read roundtrip test directory: %v", err)
	}

	for _, entry := range entries {
		name := entry.Name()
		if !strings.HasSuffix(name, ".rdn") {
			continue
		}
		baseName := strings.TrimSuffix(name, ".rdn")
		t.Run(baseName, func(t *testing.T) {
			rdnPath := filepath.Join(roundtripDir, name)
			data, err := os.ReadFile(rdnPath)
			if err != nil {
				t.Fatalf("Cannot read file: %v", err)
			}

			// Parse
			val1, err := Parse(data)
			if err != nil {
				t.Fatalf("First parse failed: %v", err)
			}

			// Stringify
			serialized, err := Stringify(val1)
			if err != nil {
				t.Fatalf("Stringify failed: %v", err)
			}

			// Parse again
			val2, err := Parse(serialized)
			if err != nil {
				t.Fatalf("Second parse failed: %v\n  serialized: %s", err, serialized)
			}

			// Compare
			if !val1.Equal(val2) {
				t.Errorf("Roundtrip mismatch:\n  original:   %s\n  serialized: %s", data, serialized)
			}
		})
	}
}
